КП АП СРНС. Этап 2

ZherebinVR opened the pull request:
Попытался реализовать алгоритм из GLONASS Satellite Coordinates Computation. Но столкнулся с полным не понимаем, как из эфемерид на определенное время получить траекторию движения спутника.

Commit:
Исправлено оформление отчета 1 части в соотвествии с ГОСТ 7.32:
Добавлено содержание, заключение.

ZherebinVR committed:
Папка source перемещена в директорию processing/

ZherebinVR committed:
Начат скрипт матлаба по вычислению координат

Korogodin commented:
Полный ответ, конечно, в ИКД, попробую описать кратко. Эфемериды ГЛОНАСС содержат координаты, компоненты вектора скорости и компоненты вектора ускорения спутника на некоторый момент времени. А дальше школьная физика - ускорения интегрируются в приращение скорости, скорость - в приращения координат. Делается это численно методом Рунге-Кутты. 

Дьявол тут в деталях - в ускорении ИКД предлагает учесть возмущение от Солнца и Луны, для чего приходится вычислять истинное звездное время и т.п. Ничего сложно, просто есть куча мест, где можно запутаться. Поэтому предлагаю действовать итеративно. 

Для начала, отбрось все ньюансы - выкинь второстепенные слагаемые из системы уравнений. Проинтегрируй, нарисуй. В итоге ты получишь ошибку в километры, но в масштабе планеты это будет незаметно (на графике SkyPlot, во времени появления спутника и т.п.). Так ты сможешь исправить крупные ошибки: с переводом МДВ в UTC, перепутанными полями в RINEX и т.п.; и напишешь всю необходимую обвязку: чтение файла, вывод графиков.  А потом уже усложняй задачу и приближайся к метровой точности.

ZherebinVR committed:
Добавлены функции расчета GMST и дифуров для численного интегрирования.
Добавлено численное интегрирование методом Рунге Кутта 4 го порядка.
Добавлена директория методических материалов /libs.

ZherebinVR committed:
Исправления численного интегрирования

ZherebinVR commented:
Понимание что и как надо делать появилось. Для численного интегрирования решил использовать функцию ode45. Как я понял, она выдает значения численного интегрирования заданной системы уравнений на определенном интервале. И вот тут началось самое интересное. 
Я записал систему уравнений:
![image](https://user-images.githubusercontent.com/55599416/79390266-14ccf180-7f78-11ea-9453-4df1da838914.png)
И в цикле решал на интервале от 0 до 1 с шагом 0.1, каждый раз подставляя новые начальные значения с предыдущего к+1 шага:
![image](https://user-images.githubusercontent.com/55599416/79390491-755c2e80-7f78-11ea-9e66-9ab53682f87b.png)
Но это решение очевидно не верное, так как результат вот такой:
![image](https://user-images.githubusercontent.com/55599416/79390570-9de42880-7f78-11ea-9a89-35e70b295707.png)
Вполне возможно, я что-то делаю абсолютно не так. Help me

Korogodin commented:
Привет! Тут всё изи. 
1. Функция diff у тебя дает правую часть системы уравнений - мгновенные значения dF производной вектора F = |X Y Z VX VY VZ|. Ты правильно положил в dF 4, 5, 6 компоненты - ускорения. А затем, зачем-то, положил в 1, 2 и 3 компонент сами координаты вместо их прозводных - компонентов вектора скорости. 
![image](https://user-images.githubusercontent.com/747285/79426873-7b7bfa80-7fcc-11ea-93e3-512e639a3399.png)

2. Вероятно, ты для начала упростил, но потом в эту функцию нужно будет ещё докинуть ускорения от Луны и Солнца. Появится зависимость от времени. 

3. Нет необходимости делать цикл самостоятельно, можно в ode сразу сгрузить вектор от t0 до конца интервала, и от t0 до начала интервала:
![image](https://user-images.githubusercontent.com/747285/79427890-11fceb80-7fce-11ea-8c79-82446c98e81e.png)

Получается как-то так
![image](https://user-images.githubusercontent.com/747285/79427971-2a6d0600-7fce-11ea-96ff-9a76bd4a6621.png)

ZherebinVR committed:
Исправление замечаний.
Добавлен пересчет координат центра масс НКА в системы координат ПЗ-90 и WGS-84.
Добавлено построение  SkyPlot.

ZherebinVR committed:
Добавлен учет ускорения от Солнца и Луны.

ZherebinVR commented:
Исправил замечания. Добавил пересчет в СК. В ПЗ-90 вышло вот так:
![image](https://user-images.githubusercontent.com/55599416/79462669-7df84780-8000-11ea-848f-1a7db16384f5.png)
А скайплот так:
![image](https://user-images.githubusercontent.com/55599416/79462715-910b1780-8000-11ea-80ab-f4846207bb78.png)
Возник вопрос про учет ускорений. Я посмотрел в ИКД, там есть примечание 1, где можно их отбросить, и потом учесть отдельно (так сделала Настя). Но в таком случае они не участвуют в системе уравнений. И зависимости от времени не появляется. Верен ли такой подход, как я сделал в diff.m?

Korogodin commented:
Вы с Настей прям аллеей одних и тех же граблей идете:
![image](https://user-images.githubusercontent.com/747285/79465184-acc3ed00-8003-11ea-889f-c164f3d1b747.png)

С ускорениями пока думать не хочу, давай хотя бы в масштабе skyplot придем к правильному результату

Korogodin commented:
И там выше ещё есть баг

![image](https://user-images.githubusercontent.com/747285/79466857-ccf4ab80-8005-11ea-8410-bb426821b686.png)

ZherebinVR commented:
А какие требования к построению Скайплот? Что является правильным результатом?
![image](https://user-images.githubusercontent.com/55599416/79469436-d6334780-8008-11ea-8b8e-33ef6fada5b5.png)
![image](https://user-images.githubusercontent.com/55599416/79469470-dfbcaf80-8008-11ea-8936-76f017fe38e4.png)

Korogodin commented:
SkyPlot - всего лишь график угла места от азимута в полярных координатах. У тебя косяк именно в расчете угла места.

ZherebinVR committed:
Исправлен баг с расчетом угла места

ZherebinVR commented:
Багуля исправлена. Странно, что она там вообще была, если этот участок кода все копировали друг у друга. Результат:
![image](https://user-images.githubusercontent.com/55599416/79472223-3d9ec680-800c-11ea-961f-65cd281809d2.png)

ZherebinVR committed:
Добавлен вариант расчета азимута для скайплота, соотвествующий Trimble

Korogodin commented:
Теперь похоже. Только ты перестарался с расписыванием арктангенса. У тебя всего лишь описка была - Z вместо X

ZherebinVR committed:
Настроен график скайплот.
Убран не правильный учет ускорения Солнца и Луны.

ZherebinVR commented:
Я настроил скайплот так, что он совпадал по азимуту с Trimble из первой части КП.
Проверил, обе формулы (через арктангенс и арккосинус) эквивалентны, выдают одинаковый результат.

Korogodin commented:
С виду всё ок, оформляй бумажки

ZherebinVR committed:
Добавлен отчет по 2 этапу.
Отчет двух этапов перенесен в корневую директорию.
Добавлен график угла места

ZherebinVR commented:
Отчет оформлен

ZherebinVR committed:
Испарвленны  опечатки.
Папка source пополнена новыми картинками и снова перемещена в корневую директорию.

Korogodin commented:
Приятное чтиво, мне понравилось, молодец)

Этап 2 принят, оценка отл
